Allow disk cache with remote exec

ron-stripe opened the pull request:
[the PR needs to have the tests cleaned up / amended, but
 I wanted to start the discussion before investing more time]

This allows usage of remote execution and a local disk cache.

The initial PR creating a combined disk and grpc cache mentioned issues with
findMissingDigests for DiskAndRemoteCacheClient,
but I noticed a couple oddities and I'm not sure if I'm missing something.
[https://cs.opensource.google/bazel/bazel/+/5f4d6995db1eb6a9d35dc163c0150283e830aa3d]

The original concern was that we should ignore the disk cache and only pay attention
to the remote cache when using DiskAndRemoteCacheClient with remote exec.  This makes
sense as bazel has to ensure the blobs are available remotely.

However, the code in findMissingDigests returns the union of all missing digests (both
disk and remote).  If it short-circuited the check by only checking  the disk cache
and only returned the disk-cache result if non-empty, I would understand the concern.
But the current code returns the union of the disk-cache result and the remote-cache result.

Additionally the current code for DiskCacheClient#findMissingBlobs unconditionally returns _all_ digests
as missing.  So in essence, the current code is always returning _all_ digests as missing.
This seems like a bug due to optimizations.
[I'm fixing this in the DiskAndRemoteCacheClient by only calling the remote for find Missing when doing remoteExec]

A test to resolve the concern would be to run remote action (which would populate both disk and remote cache).
  * verify it is in the disk cache.
  * clear the remote cache of both action cache and blobs
  * clear disk cache of blobs, but not action cache
  + run the action again.

Another concern is the current code won't upload to the remote caches (specifically the disk cache)
when the remote_exec happens.

I believe the general assumption is that most remote_exec engines will populate the remote cache themselves
so currently there isn't a call to `remoteExecutionService.uploadOutputs` when doing remote exec.
It is only there for local spawns.
Since the remote disk cache with remote exec is currently disabled, this hasn't come up.
The next time the action is attempted, it will be found in the remote_cache and pulled down.
With the current PR, the disk_cache will get populated when the ActionResult is pulled from the remote_cache
on a future run. I think that is good enough for now.

I left a note in RemoteSpawnRunner where we can update the disk cache at the
time of the RemoteExec, but that is more work to ensure we aren't doing double pushes the remote cache
so leaving out for now.

I spent some time trying to indicate in the executor log which "remote" cache client was hit:
   * disk or grpc.
Currently it is impossible to tell, but I think it would be helpful when looking a profiles, execution logs, etc.
Instead, the tests peek at the disk and remote cache (in the test tmp dir) to figure out
where actions and blobs are really stored.
If there is a way to agument DiskAndRemoteCache with more profile metrics via Profiler#profile or spawnMetricAccounting,
or ??? it would be great to add in a future PR and update the tests to use it.
Conceivably, we could use the grpc log to see which requests go to the remote cache and some fs stats to see what
goes to the disk cache :(

testing:
                                                            TEST1 disk  remote  TEST2 disk remote
  test1 -> remote_exec  [ remote_cache, disk_cache ]                 x      x
      (clear remote cache by restarting worker)                      x      -
    * verify still cached (assume remote cache)  [remote, disk]
    RESTART worker (clear cache)
    * verify still cached (assume disk cache)    [disk]              x      -

  test2 -> remote_exec  [remote_cache, disk_cache]                   x      -            x    x
     * verify
              CLEAR disk cache                                       -      -            -    x

I was hoping to post to the BuildSummaryStatsModule or ModuleCollector but can't figure out a good way.
 (any of these?)
      https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java;l=1144?q=actionResultReceived&ss=bazel%2Fbazel
        countActionResult in BuildSummaryStatsModule and MetricsCollector

coeuvre reviewed (changes requested):
Thanks for the PR!

For `findMissingDigests` in combined cache, I am thinking that should we always use the result from remote cache? Currently, there are two use cases:
  1. When uploading local inputs before remote execution, we should only use result from remote cache to make sure the missing inputs are uploaded.
  2. When uploading local outputs after local execution, if we only use result from remote cache, the worst case is outputs already in the disk cache are re-uploaded to disk cache which is, IMO, acceptable since:
     1. The chance that files in disk cache but not in remote cache is low.
     2. The cost for writing to disk compared to network IO is low.
     3. We can add existence check before writing.


When we hit the remote cache, the `runner` field of the entry in the execution log is `remote cache hit`, we can probably use `disk cache hit` for actions that hit the disk cache.

ron-stripe commented on src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java:
```diff
@@ -269,6 +269,8 @@ public SpawnResult exec(Spawn spawn, SpawnExecutionContext context)
 
             try {
+              // Don't want to reupload remote execution cache, just to disk cache
+              // remoteExecutionService.uploadOutputs(action);
```
Okay, removed the comment here.

ron-stripe commented:
> Thanks for the PR!
> 
> For `findMissingDigests` in combined cache, I am thinking that should we always use the result from remote cache? Currently, there are two use cases:
> 
> 1. When uploading local inputs before remote execution, we should only use result from remote cache to make sure the missing inputs are uploaded.
> 2. When uploading local outputs after local execution, if we only use result from remote cache, the worst case is outputs already in the disk cache are re-uploaded to disk cache which is, IMO, acceptable since:
>    
>    1. The chance that files in disk cache but not in remote cache is low.
>    2. The cost for writing to disk compared to network IO is low.
>    3. We can add existence check before writing.
> 
> When we hit the remote cache, the `runner` field of the entry in the execution log is `remote cache hit`, we can probably use `disk cache hit` for actions that hit the disk cache.

Thanks, I added the check for remote_exec in findMissing digests and only changed the behavior in that case.
Of your case 2. I think about the case of having a CI machine that starts with fresh local disk, a shared remote cache and no remote_exec.  In this case I want to ensure we still pull down to local disk.
I guess your point is that local_exec implies you aren't reading from remote_cache so you are probably right.
However, I'm not comfortable making that change.  I don't mind if you do, and I don't think it would affect the rest of the PR.

The PR grew quite large to accommodate passing spawnMetrics (builder) down to the caches so something can keep track of which cache is being used.  I was hoping to inject it rather than pass it everywhere and also keep track of the count of downloads/uploads per cache type, but I didn't see a way to do that and I wanted to limit the scope of the change.

I cleaned up the shell test and hope it is readable now.
Hopefully I'll have the other tests cleaned up and green when you see this.

coeuvre commented:
> Of your case 2. I think about the case of having a CI machine that starts with fresh local disk, a shared remote cache and no remote_exec. In this case I want to ensure we still pull down to local disk.

This this always true. `findMissingDigests` is for uploading.

> I guess your point is that local_exec implies you aren't reading from remote_cache so you are probably right.

No. local execution can read from remote cache. My point was for uploading.

> The PR grew quite large to accommodate passing spawnMetrics (builder) down to the caches so something can keep track of which cache is being used. I was hoping to inject it rather than pass it everywhere and also keep track of the count of downloads/uploads per cache type, but I didn't see a way to do that and I wanted to limit the scope of the change.

Thanks for the update! As you said, passing `spawnMetrics` down is not what we want. Another way could be updating `downloadActionResult` to return a `CachedActionResult` witch is a value class and has fields like `cacheName`, `actionResult` etc.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/common/RemoteCacheClient.java:
```diff
@@ -73,7 +73,7 @@ public int hashCode() {
    *     {@code actionKey} cannot be found the result of the Future is {@code null}.
    */
-  ListenableFuture<ActionResult> downloadActionResult(
+  FutureCachedActionResult downloadActionResult(
```
Can we return `ListenableFuture<CachedActionResult>` instead?

ron-stripe replied:
done, doing the Futures.transform was new to me ....

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/common/FutureCachedActionResult.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ */
+@Immutable
+@ThreadSafe
```
Please use `AutoValue`.

ron-stripe replied:
done

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/common/FutureCachedActionResult.java:
```diff
@@ -0,0 +1,48 @@
+// Copyright 2014 The Bazel Authors. All rights reserved.
```
Please move this to `RemoteCacheClient` just like `ActionKey`.

ron-stripe replied:
done, auto value on a nested class took a little searching :(

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -375,23 +376,29 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
+      return createFromCache(cachedActionResult, "remote");
+    }
+
+    public static RemoteActionResult createFromCache(ActionResult cachedActionResult, String cacheName) {
```
Can we change the parameter from `ActionResult` to `CachedActionResult` instead of adding another overload?

ron-stripe replied:
changed back to original way.
Passing in the CachedActionResult and storing it or pulling out the string felt awkward.
So I changed the two callers to keep the CacheActionResult and the RemoteActionResult

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -375,23 +376,29 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
 
     public RemoteActionResult(
-        ActionResult actionResult, @Nullable ExecuteResponse executeResponse) {
+        ActionResult actionResult, @Nullable ExecuteResponse executeResponse, String cacheName) {
```
Please put `cacheName` after `actionResult` since they are related.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -375,23 +376,29 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
     public static RemoteActionResult createFromResponse(ExecuteResponse response) {
       checkArgument(response.hasResult(), "response doesn't have result");
-      return new RemoteActionResult(response.getResult(), response);
+      return new RemoteActionResult(response.getResult(), response, null);
```
Please use empty string here.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteCache.java:
```diff
@@ -102,9 +103,15 @@ public RemoteCache(
   public ActionResult downloadActionResult(
```
Please change the return value to `CachedActionResult` instead of adding another download method.

ron-stripe replied:
done

ron-stripe commented:
Thanks for the comments, sorry I was pushing without realizing you would
review them all.  Part of it was letting me test on mac and linux and using
git to share.

I'll take another look and address the suggestions tomorrow.
Figuring out the right classes to use pass the cache name through the
multiple levels has been "fun".'
-ron

On Thu, Aug 19, 2021 at 8:04 PM Chi Wang ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/com/google/devtools/build/lib/remote/RemoteCache.java
> <https://github.com/bazelbuild/bazel/pull/13852#discussion_r692624715>:
>
> > @@ -102,9 +103,15 @@ public RemoteCache(
>    public ActionResult downloadActionResult(
>
> Please change the return value to CachedActionResult instead of adding
> another download method.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/13852#pullrequestreview-734625869>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARO66TBITLEBXJKQATEJQITT5XA3PANCNFSM5CGT6ZSQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

ron-stripe reviewed (commented):
Thanks, I think it is ready for another look.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/GrpcCacheClient.java:
```diff
@@ -234,9 +235,14 @@ public static boolean isRemoteCacheOptions(RemoteOptions options) {
 
-  private ListenableFuture<ActionResult> handleStatus(ListenableFuture<ActionResult> download) {
+  private ListenableFuture<CachedActionResult> handleStatus(ListenableFuture<ActionResult> download) {
+    ListenableFuture<CachedActionResult> cachedActionResultListenableFuture =
```
Can be simplified to:
```
 Futures.transform(download, CachedActionResult::remote, MoreExecutors.directExecutor());
```

ron-stripe replied:
Thanks!

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -481,19 +482,18 @@ public int hashCode() {
   /** Lookup the remote cache for the given {@link RemoteAction}. {@code null} if not found. */
   @Nullable
-  public RemoteActionResult lookupCache(RemoteAction action)
+  public CachedActionResult lookupCache(RemoteAction action)
```
Please keep this method returning `RemoteActionResult`. We may add spawn-specific types there in the future.

ron-stripe replied:
So do you want to put the cache name inside RemoteActionCache?
It felt weird to update that but okay... 
It was pretty straightforward to just create the RemoteActionCache in the two places that  called lookupCache.

I'll try and make this next round of changes tomorrow, but I'm wondering if it would be way more efficient for you take over the PR.

Or maybe @larsrc-google wants to help out for old-times-sake?

coeuvre replied:
> So do you want to put the cache name inside RemoteActionCache?

Yes. The design for `RemoteActionResult` instead of using `ActionResult` or `CachedActionResult` is we want a type to combine spawn-specific types and remote cache look up result which can be passed around late.

By taking over the PR, did you mean I create a folk of this PR and do the changes there?

ron-stripe replied:
It wasn't clear how much back and forth there would be. i.e. RemoteActionResult has 41 usages (but mostly) from tests and I wasn't sure if I should just pass in a  CachedActionResult or the name, but I think I understand your intent now.
If not, it might be close enough for you to fork and finish.  i.e. its not clear what to do with RemoteActionResult#createFromResponse.  I'll assume cacheName being null is the right thing.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:
```diff
@@ -481,19 +482,18 @@ public int hashCode() {
+        action.remoteActionExecutionContext, action.actionKey, /* inlineOutErr= */ false);
 
-    if (actionResult == null) {
+    if (cachedActionResult == null) {
```
We have lots of places checking for `cachedActionResult != null` and `cachedActionResult.actionResult() != null`. Please consider updating `downloadActionResult` to always return a future that will emit non-null `CachedActionResult` on success. We can then differentiate cache-hit and cache-miss by only checking `cachedActionResult.cacheResult() != null`.

ron-stripe replied:
done

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/disk/DiskAndRemoteCacheClient.java:
```diff
@@ -103,6 +103,14 @@ public void close() {
+    // if remote execution, find missing digests should only look at
+    // the remote cache, not the disk cache because the remote executor only
+    // has access to the remote cache, not the disk cache.
+    // Also, the current code for the DiskCache always returns all digests as missing
```
Can we re-phrase the comment so that it doesn't comparing to "current" code?

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/disk/DiskAndRemoteCacheClient.java:
```diff
@@ -182,17 +190,17 @@ private Path newTempPath() {
-          (actionResult) -> {
-            if (actionResult == null) {
+          (cachedActionResult) -> {
+            if (cachedActionResult == null || cachedActionResult.actionResult() == null) {
```
ditto.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/common/RemoteCacheClient.java:
```diff
@@ -63,6 +65,26 @@ public int hashCode() {
 
+  @AutoValue
+  abstract class CachedActionResult {
+    public static CachedActionResult create(ActionResult actionResult, String cacheName) {
```
Unless we are going to support more cache types, allowing caller to construct arbitrary cache name could lead to unintentional bugs.

Please consider only offering 4 builder methods: `remoteHit`, `diskHit`, `remoteMiss`, `diskMiss`.

ron-stripe replied:
I don't think we need the misses, but removed create.

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/common/RemoteCacheClient.java:
```diff
@@ -63,6 +65,26 @@ public int hashCode() {
+    }
+
+    @Nullable
+    public abstract ActionResult actionResult();
```
Please add java doc to describe in which case the value will be `null`.

ron-stripe replied:
Changed to be non-nullable

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/util/Utils.java:
```diff
@@ -381,16 +382,18 @@ public static String grpcAwareErrorMessage(IOException e) {
 
   @SuppressWarnings("ProtoParseWithRegistry")
-  public static ListenableFuture<ActionResult> downloadAsActionResult(
+  public static ListenableFuture<CachedActionResult> downloadAsActionResult(
```
Please keep this method returning `ListenableFuture<ActionResult>` and construct `CachedActionResult` at the call sites so we can avoid passing magic strings e.g. "disk", "remote".

coeuvre commented on src/main/java/com/google/devtools/build/lib/remote/common/RemoteCacheClient.java:
```diff
@@ -63,6 +65,26 @@ public int hashCode() {
+
+    @Nullable
+    public abstract ActionResult actionResult();
+    public abstract String cacheName();
```
Please add java doc to list the possible values and the meanings.

coeuvre reviewed (changes requested):
Thanks for the update. We are close to get it merged!

ron-stripe commented:
Thanks for the review and suggestions.
Ready for another look.

coeuvre reviewed (approved):
LGTM. I will import the PR. Thanks!

ron-stripe commented:
I've noticed this still isn't in the bazel-5 prerelease.
Is it going to go out with bazel 5.0 or 5.1?

Thanks.
-ron

```
~/bazel-5.0.0-pre.20210826.1-linux-x86_64 build --disk_cache=/tmp/fun
--config=remote  ...
ERROR: Cannot combine gRPC based remote execution with disk caching or
HTTP-based caching

On Thu, Aug 26, 2021 at 6:09 AM bazel.build machine account <
***@***.***> wrote:

> Closed #13852 <https://github.com/bazelbuild/bazel/pull/13852> via cf57d03
> <https://github.com/bazelbuild/bazel/commit/cf57d036c2e1b608ca902267fbbdfeb7ee5aa166>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/13852#event-5212830270>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARO66THGEJC6PEBGQQBW5XDT6Y4KNANCNFSM5CGT6ZSQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

brentleyjones commented:
It usually takes about 2 weeks before commits show up in one of those rolling releases.

brentleyjones commented:
FYI, this is now available since 5.0.0-pre.20210831.2.

ron-stripe commented:
Thanks I downloaded it yesterday, but I think I found a bug with a call to
`BatchReadBlobs` and am trying to track it down.

 Building src/main/java/com/stripe/build/errorprone/libutil.jar (1 source
file) failed: (Exit 34): NOT_FOUND:
BatchReadBlobs("1d6bc91d64c565d4d6321755194cb532b4000c0b06b61cb32331fc2756cc49eb/1293"):
content not found for digest
"1d6bc91d64c565d4d6321755194cb532b4000c0b06b61cb32331fc2756cc49eb/1293"

On Wed, Sep 15, 2021 at 1:49 PM Brentley Jones ***@***.***>
wrote:

> FYI, this is now available since 5.0.0-pre.20210831.2.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/13852#issuecomment-920365348>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARO66TAMWCMDLP5R2LOJXP3UCEBD3ANCNFSM5CGT6ZSQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

coeuvre commented:
IIUC, `BatchReadBlobs` is not used in Bazel yet. What's the cache setup in your case?

ron-stripe commented:
Sorry, I realized that error was from our remote build system calling
storage to get the inputs, but the inputs were missing.
I just sent another PR to fix it:
https://github.com/bazelbuild/bazel/pull/13999

bazel was thinking it didn't need to send some inputs.

On Wed, Sep 15, 2021 at 6:27 PM Chi Wang ***@***.***> wrote:

> IIUC, BatchReadBlobs is not used in Bazel yet. What's the cache setup in
> your case?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/13852#issuecomment-920504223>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARO66TGKAIUYVORJAFMI4CDUCFBX7ANCNFSM5CGT6ZSQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
